### PR TITLE
Build invocation images that run as nonroot user

### LINKED ISSUE #1921: Build bundles to run with a nonroot user
So far all bundles built with porter were built to run as the root user inside the invocation image. It is [more secure to run as a nonroot user](https://engineering.bitnami.com/articles/why-non-root-containers-are-important-for-security.html) and a requirement for inclusion of Porter in IronBank that the bundles can be run with a nonroot user.

- [ ] Build bundles that use the user nonroot by default
- [x] Inject files into the container with the correct owner and permissions 
- [ ] Evaluate if any mixins or functionality require root or additional configuration (such as any bundles that mount the docker socket)
    - [ ] Update helm3 mixin to not hard code where to find the  kubeconfig, https://github.com/MChorfa/porter-helm3/blob/f2012f4eb0ed53290e13f4a344bdd56b194d9f45/pkg/helm3/execute.go#L33
    - [ ] Ensure that mixins can use $HOME and have it resolve to /home/nonroot
    - [ ] Mixins need some build commands to run as root and others to run as nonroot
        ```Dockerfile
          ENV HELM_EXPERIMENTAL_OCI=1
          RUN apt-get update && apt-get install -y curl
          RUN curl https://get.helm.sh/helm-v3.7.0-linux-amd64.tar.gz --output helm3.tar.gz
          RUN tar -xvf helm3.tar.gz && rm helm3.tar.gz
          RUN mv linux-amd64/helm /usr/local/bin/helm3
          RUN curl -o kubectl https://storage.googleapis.com/kubernetes-release/release/v1.22.1/bin/linux/amd64/kubectl &&\
              mv kubectl /usr/local/bin && chmod a+x /usr/local/bin/kubectl
          RUN helm3 repo add bitnami https://charts.bitnami.com/bitnami
          RUN helm3 repo update
          RUN apt-get update && \
          apt-get install -y apt-transport-https curl && \
          curl -o kubectl https://storage.googleapis.com/kubernetes-release/release/v1.15.5/bin/linux/amd64/kubectl && \
          mv kubectl /usr/local/bin && \
          chmod a+x /usr/local/bin/kubectl
        ```

Depends on https://github.com/cnabio/cnab-go/pull/279

### PR DESCRIPTION
**Pull request by carolynvs**
# What does this change
As a general security principle, and also to get porter approved for inclusion in IronBank, we switched our Porter Agent to run as a nonroot user. What I didn't anticipate is that this effectively requires the invocation image (installer) to also run as the same user so that they can share files when run by the Porter Operator.

This updates porter build to run the bundle as the same nonroot user (65532) and ensure that all files in the bundle (either during build or injected before running the container) are owned by the nonroot user.

- [x] Build invocation images as nonroot
- [x] Update test bundles to not use root
- [x] Update example bundles to not use root
- [x] Update docs to not mention root
- [x] Add documentation page about how we run as nonroot (need security page) Created follow up #1944
- [x] Add note to the custom dockerfile page about updating your template with new changes to porter.
- [x] Get it working with allow-docker-host-access
- [x] Test with docker and docker-compose mixins

This commit message has a ton of context into why certain changes were made to file permissions, how we mount the docker socket, etc. https://github.com/getporter/porter/pull/1930/commits/bfefe60082edcf7ce6ea3203257192bb23b9f73f

I will squash when I merge, sorry this is such a giant, awkward PR!

# What issue does it fix
Closes #1921 

# Notes for the reviewer
Depends on https://github.com/MChorfa/porter-helm3/pull/42

# Checklist
- [x] Did you write tests?
- [x] Did you write documentation?
- [ ] Did you change porter.yaml or a storage document record? Update the corresponding schema file.
- [ ] If this is your first pull request, please add your name to the bottom of our [Contributors][contributors] list. Thank you for making Porter better! 🙇‍♀️

# Reviewer Checklist
* Comment with /azp run test-porter-release if a magefile or build script was modified
* Comment with /azp run porter-integration if it's a non-trivial PR

[contributors]: https://porter.sh/src/CONTRIBUTORS.md

### CODE REVIEW

File: pkg/build/dockerfile-generator_test.go [hunk 1]
```diff
@@ -46,98 +44,11 @@ func TestPorter_buildDockerfile(t *testing.T) {
...
 			wantDockerfilePath := fmt.Sprintf("testdata/%s.Dockerfile", driver)
-			wantDockerfile, err := ioutil.ReadFile(wantDockerfilePath)
-			require.NoError(t, err, "could not read %s", wantDockerfilePath)
-			assert.Equal(t, string(wantDockerfile), gotDockerfile)
+			test.CompareGoldenFile(t, wantDockerfilePath, gotDockerfile)
 		})
 	}
 }
 
-func TestPorter_buildDockerfile_alternateManifestLocation(t *testing.T) {
```
Comment (carolynvs) [inline pkg/build/dockerfile-generator_test.go:56]: I removed some tests that are no longer relevant or duplicate validations from other tests

File: tests/smoke/copy_test.go [hunk 2]
```diff
@@ -21,6 +20,10 @@ import (
 // Stop the temporary registry
 // Copy the bundle to another location, this will fail unless we are properly using the relocation map
 func TestCopy(t *testing.T) {
+	test, err := tester.NewTest(t)
```
Comment (carolynvs) [inline tests/smoke/copy_test.go:23]: This test was failing and preventing me from merging this PR (even though it's unrelated) because it's a flake #1928

**Comment by VinozzZ** [comment]
It looks great to me. Jus one thing that I ran into while testing the change with --allow-docker-host-access flag. It seems like the docker.socket isn't accessible to the nonroot user in the container since the socket is created by root. I tried to change permission for it in the invocation image but that didn't work.
Maybe we need some logic when the flag is specified to not use nonroot containers?

**Comment by carolynvs** [comment]
@VinozzZ  Thanks for the reminder! I completely forgot to test out --allow-docker-host-access. Since you had so much trouble getting it to work with a nonroot user, I'll try for a bit to get it working but then may switch bundles that use that feature back to root. Those types of bundles (that directly access the docker socket) don't work with kubernetes anyway so running as a different user isn't a concern.

File: pkg/encoding/encoding.go [hunk 3]
```diff
@@ -26,7 +27,7 @@ func MarshalFile(fs aferox.Aferox, path string, in interface{}) error {
 	if err != nil {
 		return err
 	}
-	return fs.WriteFile(path, data, 0700)
+	return fs.WriteFile(path, data, pkg.FileModeWritable)
```
Comment (carolynvs) [inline pkg/encoding/encoding.go:30]: This never should have been set with +x, so I'm flipping it to just writable

File: pkg/build/build.go [hunk 4]
```diff
@@ -33,8 +33,13 @@ var (
...
 	BUNDLE_DIR = "/cnab/app"
 
-	// INJECT_PORTER_MIXINS_TOKEN can control where mixin instructions will be placed in Dockerfile.
-	INJECT_PORTER_MIXINS_TOKEN = "# PORTER_MIXINS"
+	// PORTER_MIXINS_TOKEN can control where mixin instructions will be placed in
+	// Dockerfile.
+	PORTER_MIXINS_TOKEN = "# PORTER_MIXINS"
+
+	// PORTER_INIT_TOKEN controls where Porter's image initialization
+	// instructions are places in the Dockerfile.
```
Comment (vdice) [inline pkg/build/build.go:41]:
```suggestion
	// instructions are placed in the Dockerfile.
```

**Review by vdice — APPROVED** [review]
Awesome work, both here and in cnab-go!  👏 There were a lot of spots to attend to. LGTM.

**Pull request merged**